clients/src/main/java/org/apache/kafka/clients/consumer/NoOffsetForPartitionException.java

```diff
@@ -26,19 +26,20 @@
 
     private static final long serialVersionUID = 1L;
 
-    private final TopicPartition partition;
+    private final Set<TopicPartition> partitionSet;
 
-    public NoOffsetForPartitionException(TopicPartition partition) {
-        super("Undefined offset with no reset policy for partition: " + partition);
-        this.partition = partition;
+    public NoOffsetForPartitionException(String message) {
```

Comment:
I dunno... I guess we could have bundled up more information in the return value of `poll()`, but I think that would have just added more api overhead for users who are probably mostly using a defined reset policy anyway (in which case we handle this error internally). We also considered adding a callback which would be invoked on this error, but again, that doesn't actually seem like a simpler API. In this case, I think throwing the exception is a reasonable middle ground, especially since this case should be rare even for users who have no reset policy. In other words, I think it really is an exceptional case.

I agree that the current usage is a little poor though. It would be better to get the full set of partitions that have no offsets before throwing. This would be consistent with the way we handle other offset-related errors. It's also the same pattern forTopicAuthorizationException, which returns the full set of unauthorized topics (which may be a subset of the subscribed topics). If we don't provide any details on which topics are unauthorized, then the user has to figure it out themselves, which would be annoying.

Anyway, I'm not sure what the answer is here, but I think we should make the exceptions we throw useful. In my opinion, that means reliably including whatever information is needed to handle them (even if users just log the errors, the extra information helps in debugging). Perhaps we need a new class of user exceptions which is separate from api exceptions? Or maybe the static mapping from error code to exception instance is too restrictive in `Errors` and we need some way to pass an error context?